### DISCUSSION TITLE
Does sktime support panel forecasting with pandas MultiIndex?

### DISCUSSION THREAD

**Discussion by roboswell** [alan-turing-institute/sktime#1271] [Q&A]
I have country-month data, and have MultiIndexed it so that forecasting can ideally keep track of predictions in terms of which countries are in my dataset - not simply when an event is predicted to happen "somewhere in the world" at a specific time. 

When I run the code:

forecaster1 = EnsembleForecaster([
    ("logit", LogisticRegression(penalty='none', class_weight='balanced', random_state=36))
])

cv = ExpandingWindowSplitter(
    step_length=1, fh=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], initial_window=84
)

df = evaluate(forecaster=forecaster1, X=X, y=y, cv=cv, strategy="refit", return_data=True)
 
The error I get is: 

NotImplementedError: <class 'pandas.core.indexes.multi.MultiIndex'> is not supported, use one of (<class 'pandas.core.indexes.numeric.Int64Index'>, <class 'pandas.core.indexes.range.RangeIndex'>, <class 'pandas.core.indexes.period.PeriodIndex'>, <class 'pandas.core.indexes.datetimes.DatetimeIndex'>) instead.

Is there a way to index BOTH time and country with sktime? I'm fairly new to this. Thanks

**Comment by fkiraly** [comment]
The task you are describing is called panel foreasting, because you have variables indexed by country.

You can in-principle use `sktime` for this already, by converting your panel into wide format, i.e., having a row index which is time, and columns which area country/variable multi-index (or simply a string index).
Though dedicated panel forecasting methods are not yet implemented - contributions are very much appreciated.

**Reply by fkiraly** [reply]
see here for an issue on a specific method https://github.com/alan-turing-institute/sktime/issues/1132
see here for an extension proposal related to this https://github.com/sktime/enhancement-proposals/pull/9

**Comment by fkiraly** [comment]
FYI @roboswell, there is now a workstream for this: https://github.com/alan-turing-institute/sktime/projects/25

**Comment by fkiraly** [comment]
@roboswell, the `sktime` base class interface now supports panel (and hierarchical) forecasting with `pandas.MultiIndex`.
Currently (at 0.10.1) the changes are on the development `main` branch and not yet in a release, see the closing message #1272 and the PR #1818. The next release version, 0.11.0, should contain this functionality.

If you are still interested in panel forecasting functionality, we would appreciate if you try it out and play around with it.
Implementing additional panel or hierarchical foreceasters should also be (hopefully) easy now.

**Comment by fkiraly** [comment]
@roboswell, this is now part of `sktime` base functionality. All forecasters either suppor tthis natively, or by vectorizing/looping over panel instances.

**Comment by vroomzel** [comment]
The main forecasting tutorial mentions that more details is available in the hierarchical forecasting tutorial, but i wasn't able to find that one. Does hierarchical/panel forecasting tutorial exists?

**Reply by fkiraly** [reply]
@vroomzel, yes, it exists!
It is in the repository for the pydata 2022 Berlin presentation, perhaps we should move it to the main repo:
https://github.com/sktime/sktime-workshop-pydata-london-2022

**Reply by vroomzel** [reply]
Found it. Thank you!!
https://github.com/sktime/sktime-tutorial-pydata-berlin-2022/blob/main/notebooks/3_hierarchical_global_forecasting.ipynb
